srt-protocol/src/listener/mod.rs

```diff
@@ -0,0 +1,390 @@
+mod input;
+mod session;
+mod statistics;
+
+use std::{collections::HashMap, fmt::Debug, net::SocketAddr, time::Duration, time::Instant};
+
+use crate::{packet::*, protocol::time::Timer, settings::ConnInitSettings};
+
+use session::*;
+
+pub use crate::protocol::pending_connection::{AccessControlRequest, AccessControlResponse};
+pub use input::*;
+pub use statistics::*;
+
+#[derive(Clone, Debug)]
+pub struct ListenerSettings {}
+
+#[derive(Debug)]
+pub struct MultiplexListener {
+    start_time: Instant,
+    local_address: SocketAddr,
+    settings: ConnInitSettings,
+    sessions: HashMap<SessionId, SessionState>,
+    stats: ListenerStatistics,
+    stats_timer: Timer,
+}
+
+impl MultiplexListener {
+    pub fn new(now: Instant, local_address: SocketAddr, settings: ConnInitSettings) -> Self {
+        Self {
+            start_time: now,
+            local_address,
+            settings,
+            sessions: Default::default(),
+            stats: Default::default(),
+            stats_timer: Timer::new(now, Duration::from_secs(1)),
+        }
+    }
+
+    pub fn handle_input<'s, 'a>(&'s mut self, now: Instant, input: Input) -> Action<'a>
+    where
+        'a: 's,
+    {
+        self.debug(now, "input", &input);
+
+        let action = match input {
+            Input::Packet(packet) => self.handle_input_packet(now, packet),
+            Input::AccessResponse(response) => self.handle_input_access_response(now, response),
+            Input::Timer => self.handle_timer(now),
+            Input::Success(result_of) => self.handle_success(now, result_of),
+            Input::Failure(result_of) => self.handle_failure(now, result_of),
+        };
+
+        self.debug(now, "action", &action);
+
+        action
+    }
+
+    fn handle_input_packet<'s, 'a>(
+        &'s mut self,
+        now: Instant,
+        packet: ReceivePacketResult,
+    ) -> Action<'a>
+    where
+        'a: 's,
+    {
+        match packet {
+            Ok(packet) => self.handle_packet(now, packet),
+            Err(error) => self.handle_packet_receive_error(now, error),
+        }
+    }
+
+    fn handle_input_access_response<'s, 'a>(
+        &'s mut self,
+        now: Instant,
+        response: Option<(SessionId, AccessControlResponse)>,
+    ) -> Action<'a>
+    where
+        'a: 's,
+    {
+        match response {
+            Some((session_id, response)) => {
+                self.handle_access_control_response(now, session_id, response)
+            }
+            None => self.handle_close(),
+        }
+    }
+
+    fn handle_packet<'s, 'a>(&'s mut self, now: Instant, packet: (Packet, SocketAddr)) -> Action<'a>
+    where
+        'a: 's,
+    {
+        self.stats.rx_packets += 1;
+        //self.stats.rx_bytes += packet
+        let session_id = SessionId(packet.1, packet.0.dest_sockid());
+        let session = match self.sessions.get_mut(&session_id) {
```

Comment:
This should use the Entry api, will be cleaner and more performant

Reply:
good point, I haven't used HashMap much yet